internal/research/opentelemetry/instrumentation-metrics.md

```diff
@@ -0,0 +1,43 @@
+# OpenTelemetry SDK Usage for Metrics in Golang Applications
+
+OpenTelemetry is a project to handle telemetry data (logs, metrics and traces) in a vendor-neutral way. It provides SDKs to instrument applications, APIs and processing tools (collector).
+
+This document describes our findings using the OpenTelemetry Golang SDK for metrics export in a microservice. To evaluate the functionality of the SDK, a Golang implementation for the reviews service of the [Istio Bookinfo](https://istio.io/latest/docs/examples/bookinfo/) example was extended.
+
+## Metric Types
+
+OpenTelemetry supports metrics of type `counter` (increasing value) , `observer` (gauge value), and `measure`. Measures can be combined using aggregations. A possible aggregation is a histogram. The metric types are represented by the `Int64Counter`, `Int64UpDownCounter`, and `Int64ValueRecorder` types in the Golang SDK.
+
+## Code Instrumentation
+
+To instrument a Golang application, first, a `Meter` object must be created. To use the use the Prometheus protocol, it has to be exposed as an http handler:
+
+```
+exporter, _ := prometheus.InstallNewPipeline(prometheus.Config{})
+http.HandleFunc("/", exporter.ServeHTTP)
+meter := global.Meter("my-service")
+```
+
+The exporter initialization is identical to the [tracing instrumentation](instrumentation-tracing.md) for the OTLP protocol.
+
+Individual code snippets can be instrumented with metrics by first creating a metric:
+
+```
+requestCounter := metric.Must(meter).NewInt64Counter("http.server.requests.total", metric.WithDescription("measures total the number of HTTP requests"))
+```
+
+And modified for example in a HTTP request handler:
+
+```
+h.requestCounter.Add(r.Context(), 1)
+```
+
+## Findings
+
+* The metrics API is declared to be unstable and has no documentation yet.
+* Metric types in the Golang SDK abstract from the wire protocol. For instance `measure` vs `Int64ValueRecorder`.
+* Access to histograms that are derived from a `Int64ValueRecorder` metric do via Grafana does not show any values.
+
+## Conclusions
+
+OpenTelemetry provides similar capabilities as the [Prometheus Golang client library](https://github.com/prometheus/client_golang) for the supported metric types and intrumentation steps. However, the early stage and expected changes of the SDK does not make it suitable for production use.
```

Comment:
I'd also mention the following capabilities of the Prometheus client library that OpenTelemetry SDK for Go is lacking:
1. Automatic collection of Go runtime metrics/process metrics.
2. Middleware for request handler instrumentation (https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/promhttp)

Reply:
Done